### PR TITLE
Add public registration

### LINKED ISSUE #21981: Implement Public User Registration
Implement feature request https://github.com/directus/directus/discussions/3083 by allowing public user registration.

Validations to enforce during the creation and update of `users`.

### Conditions: Admin access, App access,  API access
  - Admin User (Admin & App)
  - App User(App)
  - API User

### Suggested limits:
  - `USERS_MAX_ADMIN_ACCESS`
      - Total number of Admin users
  - `USERS_MAX_APP_ACCESS`
      - Admin + App Access total count
  - `USERS_MAX_API_ACCESS`
      - Total users including Admin + App Access

### PR DESCRIPTION
**Pull request by DanielBiegler**
## Scope

What's changed:

- New UI so you can go to register
- It adds a new user with a specified role
- Fixed a UI bug on the login page
- DB migration that adds fields to directus_settings
- Hide register button depending on directus_settings
- Email validation via customizable filter
- Add mailing service to send out the verification code
- Email template for the verification email
- New per-ip-rate-limiter for registration, their env vars and docs
- Remove `RATE_LIMITER_GLOBAL_STORE` both from code and docs since we dont use it. Lets stick to one store for all limiters.

| Settings | Email |
|--------|--------|
| image | image | 

| Register Route | After Success |
|--------|--------|
| image | image | 

- Fixed UI inconsistency in login & reset-password

| Before | After |
|--------|--------|
| <video src="https://github.com/directus/directus/assets/14810858/13b888dc-dd3a-44ba-aeae-ed09aa949e60" autoplay loop muted></video> | <video autoplay loop muted src="https://github.com/directus/directus/assets/14810858/3d01ef66-041a-4998-ba7b-626eda9ad750"></video> | 

## Potential Risks / Drawbacks

- todo

## Review Notes / Questions

- Dont forget to run the migration :wink: 

---

Fixes #21981

### CODE REVIEW

File: api/src/services/users.ts
Comment (DanielBiegler) [inline api/src/services/users.ts:494]: I'm thinking about renaming the `invited`-status to a more generic `pending`-status so that we can use the same status for invited and registered-but-not-yet-verified users but simply using the migration file to rename the status could lead to breaking existing logic in users projects for example somebody might have Flows that expect the users to be `invited`

Another approach could be to just replace the `Invited` UI label to `Pending` and keep the underlying `invited` string the same to stay backwards compatible. 

Mhh. Could also just add a new status to be both precise and backwards compatible. Like so:

image

If we add the new status we can give it a more concrete name too like `Unverified` for example

File: app/src/routes/register/register.vue
Comment (DanielBiegler) [inline app/src/routes/register/register.vue:65]: actually, do we want to show sso links on the register page? :thinking:

File: api/src/controllers/users.ts
Comment (DanielBiegler) [inline api/src/controllers/users.ts:510]: It'd be nice if we would add a way to specify more user fields because I can see people using this service to build their own login pages with more fields :thinking: 

Allowing other user fields would need to be sanitized though we cant let them override role for example

File: sdk/src/rest/commands/server/info.ts
Comment (paescuj) [inline sdk/src/rest/commands/server/info.ts:8]: This should be sufficient IMO, the type already indicates that it's a toggle state:
```suggestion
		public_registration: boolean;
```

E.g. comparable to states in Collection metadata:
https://github.com/directus/directus/blob/d10385d30305adbe1b9abfa4bdefcff1836effca/packages/system-data/src/types.ts#L26-L33

File: app/src/routes/register/register.vue
Comment (br41nslug) [reply app/src/routes/register/register.vue:65]: > as those fall under the action of "login", right

Not necessarily if you have public registration enabled for SSO it doubles as both login and registration 🤔 
I wouldnt be against having the SSO buttons on the registration page if configured as such but feels like someone with more UX insight may have more opinions there.

File: api/src/services/users.ts
Comment (paescuj) [reply api/src/services/users.ts:494]: I would have to take a closer look, but I think a dedicated (hidden) field for the email verification status could make sense. This could also be used, for example, if the user is already active and changes the email address, which would have to be verified again (but the user status should stay "active").

File: app/src/routes/register/register.vue
Comment (DanielBiegler) [reply app/src/routes/register/register.vue:65]: Agreed, removing them now

File: api/src/controllers/users.ts
Comment (paescuj) [reply api/src/controllers/users.ts:510]: You mean like require a phone number or similar? 🤔

Comment (DanielBiegler) [reply api/src/controllers/users.ts:510]: Yes for example a number or even simpler first & last name. Avatar could also be useful, stuff like that. It'd be nice to also allow populating user-provided fields for when users have manually extended the `directus_users` table so it'd have to be dynamic but I'm a little hesitant there because of security concerns. Mhh

File: app/src/routes/register/register.vue
Comment (paescuj) [reply app/src/routes/register/register.vue]: > Not necessarily if you have public registration enabled for SSO it doubles as both login and registration 🤔

Yeah, but you actually login to the provider with an existing account there (at least in most cases)

File: api/src/controllers/users.ts
Comment (paescuj) [reply api/src/controllers/users.ts:510]: Interesting ❤️ Well, actually it would "just" be another (public/shared) form where we could use existing validation logic, etc. 🤔

File: app/src/lang/translations/en-US.yaml
Comment (hanneskuettner) [inline app/src/lang/translations/en-US.yaml:1368]: These should probably both be title case 

```suggestion
    public_registration: Public Registration
    is_public_registration_enabled: Public Registration
```

File: app/src/routes/register/register-form.vue
Comment (DanielBiegler) [reply app/src/routes/register/register-form.vue:47]: This made me check and turns out `@directus/errors` is currently not a dependency of the app?! :exploding_head:  Interesting. Good find, thanks.

File: .changeset/dry-otters-attend.md
Comment (licitdev) [inline .changeset/dry-otters-attend.md:2]: Should this be a `minor` release to increment the `directus` version?

```suggestion
'directus': minor
'@directus/system-data': patch
```

File: api/src/database/migrations/20240422A-public-registration.ts
Comment (paescuj) [inline api/src/database/migrations/20240422A-public-registration.ts:3]: Bit of a nitpick, but since this migration is fairly small, I'd prefer to inline these values which would improve the readability IMO. That's also how we're doing it in the existing migrations.

File: api/src/services/mail/templates/user-registration.liquid
Comment (hanneskuettner) [inline api/src/services/mail/templates/user-registration.liquid:8]: Should we add one of those standard disclaimers a la "If you don’t want to create an account, you can ignore this message."

File: api/src/services/users.ts
Comment (hanneskuettner) [inline api/src/services/users.ts:521]: Should probably be title case? That's at least that's what all my emails I can see in my inbox have.

File: packages/system-data/src/fields/settings.yaml
Comment (hanneskuettner) [inline packages/system-data/src/fields/settings.yaml:225]: Do we have the possibility to add validation rules here? That the user can not leave this `null` if the public registration is turned on? Don't know if that is possible in there definition files, but would be a nice to have.

File: packages/system-data/src/fields/settings.yaml
Comment (hanneskuettner) [inline packages/system-data/src/fields/settings.yaml:237]: Personally I'd swap verify email and the email filter, since the email filter is more closely related to the user role (conceptionally) and the send verification email more related to a process of the registration as a whole.

**Comment by hanneskuettner** [comment]
One whole new thought that came to me while reviewing this. Should there be an option to avoid creating any users until the user clicks the verification link? This would fully avoid having the possibility of filling a DB with bogus users.

Theoretically we can pass the full user payload through the signed token (which would not allow any tempering) and delay the account creation until the verification callback.

File: packages/system-data/src/fields/settings.yaml
Comment (hanneskuettner) [inline packages/system-data/src/fields/settings.yaml:1]: I think we talked about it, but I feel like the options for the registration should be disabled if the registration feature itself is disabled.

File: app/src/routes/register/register.vue
Comment (hanneskuettner) [inline app/src/routes/register/register.vue:34]: Does it make sense to have two different notices depending on wether the verification email is actually enabled or not. UX wise it feels a bit weird and I could see users waiting for an email to arrive with the notice being worded like this.

File: packages/env/src/constants/defaults.ts
Comment (hanneskuettner) [inline packages/env/src/constants/defaults.ts:33]: Should this be enabled by default since we're basically handing the users a loaded gun here?

File: packages/system-data/src/fields/settings.yaml
Comment (br41nslug) [reply packages/system-data/src/fields/settings.yaml:1]: In that same vein do we disable local registration when local auth is disabled? `AUTH_DISABLE_DEFAULT`

File: app/src/lang/translations/en-US.yaml
Comment (hanneskuettner) [inline app/src/lang/translations/en-US.yaml:1368]: Maybe we can use technical wording? 

```suggestion
    public_registration_note: Allows users to register via a [registration page](/admin/register).
```

File: packages/env/src/constants/defaults.ts
Comment (DanielBiegler) [reply packages/env/src/constants/defaults.ts:33]: Yeah thought the same thing, but we dont enable other rate limiters by default either. Both the normal and global one are disabled by default too. Mhh :thinking:

File: packages/system-data/src/fields/settings.yaml
Comment (hanneskuettner) [reply packages/system-data/src/fields/settings.yaml:1]: Yes! And they can make changes without saving the page, can they not?

File: app/src/routes/register/register-form.vue
Comment (DanielBiegler) [reply app/src/routes/register/register-form.vue:83]: The consensus was that the toggleable password input would be nice-to-have and preferred but so far I havent gotten around to it. As far as I'm aware, we would need a new component for that. :+1:

File: api/src/services/mail/templates/user-registration.liquid
Comment (DanielBiegler) [reply api/src/services/mail/templates/user-registration.liquid:8]: Yeah, sounds good, agreed! :+1:

File: packages/env/src/constants/defaults.ts
Comment (paescuj) [reply packages/env/src/constants/defaults.ts:33]: This is to be compared to the login rate limiter which cannot be turned off either 👍

File: app/src/lang/translations/en-US.yaml
Comment (paescuj) [reply app/src/lang/translations/en-US.yaml:1368]: Additionally, I'd probably leave out the link. Maybe something like this?
```suggestion
    public_registration_note: Allows users to self-register to the project.
```

File: api/src/services/mail/templates/user-registration.liquid
Comment (paescuj) [reply api/src/services/mail/templates/user-registration.liquid:6]: Or even
```suggestion
  You have registered at <i>{{ projectName }}</i>.
```

File: app/src/routes/register/register.vue
Comment (hanneskuettner) [inline app/src/routes/register/register.vue:1]: IMHO this route should have a `beforeEnter` guard that redirects logged in users to either `/login` or directly to their last page. It feels unfinished if a user can view the register page as a logged in user.

File: app/src/lang/translations/en-US.yaml
Comment (DanielBiegler) [reply app/src/lang/translations/en-US.yaml:1368]: Linking is useful because not everyone will know where the registration will appear or how. Why would less information here be better than linking to the page?

File: app/src/routes/register/register.vue
Comment (licitdev) [reply app/src/routes/register/register.vue:1]: +1. 👍 This goes with Pascal's comment to remove the link to this page. https://github.com/directus/directus/pull/22125/files#r1591511293

Comment (DanielBiegler) [reply app/src/routes/register/register.vue:1]: If theres no technical reason for forbidding logged in users reaching that page, I dont see the value. Under normal usage for normal users, you cant even see the link to that route on the log-in page because of the "you are currently logged in as"-component. So non-admins would have to manually type in that url or log out. And at that point just let people go there. Admins will see the link in the project settings but normal users dont. Without a technical reason this seems unimportant to me. If we decide that we need to forbid api usage of that route while someone is logged in, then yes, we should also redirect them but thats currently not the case.

File: packages/system-data/src/fields/settings.yaml
Comment (DanielBiegler) [reply packages/system-data/src/fields/settings.yaml:1]: > And they can make changes without saving the page, can they not?

Not sure if I understand you here? If I wanna just update the email filter I would then have to enable the registration, update the email filter, disable the registration again and then save. That doesnt sound useful to me

File: app/src/lang/translations/en-US.yaml
Comment (DanielBiegler) [reply app/src/lang/translations/en-US.yaml:1374]: I had send verification email beforehand and Ben mentioned in the show&tell we can shorten it to verify email. Leaving it for now. We can change it later if necessary :+1:

File: packages/system-data/src/fields/settings.yaml
Comment (DanielBiegler) [reply packages/system-data/src/fields/settings.yaml:237]: I prefer the toggles being on the left side while the more configurable fields live on the right. Since this is just an opinion thing I'd leave it up to @benhaynes / @rijkvanzanten - For quick reference here is how it currently looks like:

image

File: app/src/lang/translations/en-US.yaml
Comment (paescuj) [reply app/src/lang/translations/en-US.yaml:1368]: The feature won't be limited to the register page of the Data Studio, but also for registering via API, so a more general note seems appropriate to me 👍

File: packages/env/src/constants/defaults.ts
Comment (DanielBiegler) [reply packages/env/src/constants/defaults.ts:33]: Not every environment cares about limiting this so I'd keep the ability to turn it off but maybe we indeed should enable it by default with conservative defaults

File: app/src/routes/register/register.vue
Comment (licitdev) [reply app/src/routes/register/register.vue:1]: I rest my case. 👍  Just noticed that the `/reset-password` route works the same way as `/register`.

File: api/src/services/users.ts
Comment (DanielBiegler) [reply api/src/services/users.ts:512]: Mhh initially I agreed because I thought we can reuse it for more tokens but now I looked up our other TTLs and theres no further overlap, just the invite and registration. Furthermore it sounds like this could be a nice configuration env variable even! For example we also have `SESSION_COOKIE_TTL`, `ACCESS_TOKEN_TTL` so why not add something like `EMAIL_VERIFICATION_TOKEN_TTL`, I can already see other people wanting to override this to shorter time frames. I have not thought about this initially, thanks! :heart:  :+1:

File: app/src/routes/register/register.vue
Comment (paescuj) [reply app/src/routes/register/register.vue:1]: Generally, I agree that there is no compelling reason to hide this page for logged-in users.
However, what's a bit confusing at the moment, is that when you register while already logged-in, the success message appears, with the "sign in" button, even though you are (obviously) already signed-in as current user - maybe a "switch user" button or similar would make that clearer...

On a similar note, the register page is currently available, even when public registration is disabled - but I guess that's fine too.

File: api/src/database/migrations/20240422A-public-registration.ts
Comment (DanielBiegler) [reply api/src/database/migrations/20240422A-public-registration.ts:3]: If it _strongly_ irks you or someone else, we can change it, but personally I dont mind the variables ¯\\\_(ツ)\_/¯

File: app/src/lang/translations/en-US.yaml
Comment (DanielBiegler) [reply app/src/lang/translations/en-US.yaml:1368]: > The feature won't be limited to the register page

Ahh true, good point! But I still like linking to it, I regard that as useful UX for admins. I used it quite often. mhh

File: api/src/database/migrations/20240422A-public-registration.ts
Comment (DanielBiegler) [reply api/src/database/migrations/20240422A-public-registration.ts]: Done

File: api/src/services/users.ts
Comment (licitdev) [reply api/src/services/users.ts:507]: Yes, the user can still click on the verification link to activate their account if the token is still valid. It's an edge case for users in the pending verification `draft` status with expired tokens where the email verification requirement was disabled after their registration.

The logical check in the preceding block ensures that `user.status === 'draft'`. 
https://github.com/directus/directus/blob/854dd1c1d7b67d560a5a6b19a8fb824300efdc9e/api/src/services/users.ts#L503-L507

I can't think of a scenario right now where it backfires. 🤔

**Review by licitdev — COMMENTED** [review]
Will the `graphql` and `sdk` commands for public user registration be included in this PR?

File: api/src/services/users.ts
Comment (hanneskuettner) [reply api/src/services/users.ts:507]: Would an edge case be, similar to the one you mentioned below, where this could result in a 'draft' admin account being activated through this flow?

Comment (hanneskuettner) [reply api/src/services/users.ts:507]: Which already is a problem of the current version, tbh

Comment (licitdev) [reply api/src/services/users.ts:507]: Yes, they are kind of related as both involve existing `draft` accounts. However, with the introduction of a new status, this edge case remains.

The logical block linked above could be updated with the new `status`, but the code suggestion remains.

Comment (hanneskuettner) [reply api/src/services/users.ts:507]: So the "attack" would be that a new admin user has been set up, but is still intended to by in the `draft` status until another admin actives the account. 
That user can, if public registration is enabled and email verification disabled self activate their account, reset their password, as they now have an active account and then log in to the system, even though that user was never admitted as an active admin.
That seems unlikely, but still a valid edge case that should not be possible IMO.

File: api/src/middleware/rate-limiter-registration.ts
Comment (br41nslug) [reply api/src/middleware/rate-limiter-registration.ts:1]: We can abstract this in a later optimization if need be.

File: api/src/services/users.ts
Comment (DanielBiegler) [reply api/src/services/users.ts]: Done :)

File: app/src/routes/register/register-form.vue
Comment (hanneskuettner) [inline app/src/routes/register/register-form.vue:82]: Nitpick: Add `autocomplete="new-password"` to play nice with browsers and password managers.

```suggestion
		
		<v-input
			v-model="password"
			type="password"
			autocomplete="new-password"
			:placeholder="t('password')"
			:disabled="isLoading"
		/>
```

File: app/src/routes/register/register-form.vue
Comment (hanneskuettner) [inline app/src/routes/register/register-form.vue:83]: Same as above

```suggestion
		<v-input
			v-model="passwordVerification"
			type="password"
			autocomplete="new-password"
			:placeholder="t('password')"
			:disabled="isLoading"
		/>
```

File: app/src/routes/register/register-form.vue
Comment (br41nslug) [reply app/src/routes/register/register-form.vue:82]: nice one 😄

File: app/src/lang/translations/en-US.yaml
Comment (paescuj) [reply app/src/lang/translations/en-US.yaml:1369]: Maybe a combination of both then?
```suggestion
    public_registration_note: Allows users to register via the [registration page](/admin/register) and API.
```

```suggestion
    public_registration_note: Allows users to self-register via the [registration page](/admin/register) and API.
```

File: app/src/lang/translations/en-US.yaml
Comment (paescuj) [reply app/src/lang/translations/en-US.yaml:1377]:
```suggestion
      Send an email to newly signed-up users to verify their email address before they can sign in.
```

File: app/src/lang/translations/en-US.yaml
Comment (paescuj) [inline app/src/lang/translations/en-US.yaml:1651]: Translation keys should be somewhat neutral (and existing ones generally are)
```suggestion
no_account: Don't have an account?
already_an_account: Already have an account?
```

File: app/src/routes/register/register-form.vue
Comment (paescuj) [inline app/src/routes/register/register-form.vue:22]: AFAICS, we only seem to care about the event itself
```suggestion
	success: [];
```

File: app/src/routes/register/register-form.vue
Comment (paescuj) [inline app/src/routes/register/register-form.vue:29]: This message makes sense for login, but I think for registering it should be: Invalid username or password.

File: app/src/routes/register/register-form.vue
Comment (paescuj) [inline app/src/routes/register/register-form.vue:65]: As mentioned above
```suggestion
		emit('success');
	} catch (err: any) {
		error.value = err.errors?.[0]?.extensions?.code || err;
```

File: app/src/routes/register/register.vue
Comment (paescuj) [inline app/src/routes/register/register.vue:45]:
```suggestion
		<register-form v-else @success="wasSuccessful = true" />
```

File: api/src/controllers/users.ts
Comment (paescuj) [inline api/src/controllers/users.ts:530]: Wondering if this should just be `/register/verify`, similar to other single-worded routes like `/invite/accept`, etc. and also to be in-line with the service method name, SDK command, ...

**Review by paescuj — COMMENTED** [review]
- What terminology have we agreed on? In the Data Studio "User Registration" is used, while the API configs use "public_registration". This should be aligned before merging.
- Docs are missing for the new Project Settings [^1], not sure if we still want to maintain this page, though.

[^1]: https://docs.directus.io/user-guide/settings/project-settings.html

**Review by br41nslug — APPROVED** [review]
🚀

File: api/src/services/server.ts
Comment (DanielBiegler) [reply api/src/services/server.ts:90]: Good point, on it

**Pull request merged**
